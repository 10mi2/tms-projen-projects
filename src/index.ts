export * from "./tms-typescript-app-project";
export * from "./tms-nestjs-app-project";
