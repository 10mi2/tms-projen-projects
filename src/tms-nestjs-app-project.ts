import * as path from "node:path";
import { SampleDir, SampleFile, javascript } from "projen";
import { deepMerge } from "projen/lib/util";
import {
  TmsTypeScriptAppProject,
  TmsTypeScriptAppProjectOptions,
} from "./tms-typescript-app-project";

export interface TmsNestJSAppProjectOptions
  extends TmsTypeScriptAppProjectOptions {
  /**
   * Which type of sample code to include, if `sampleCode` is true.
   *
   * @default "graphql-codefirst"
   * @featured
   */
  readonly sampleType?: "graphql-codefirst" | "graphql-schemafirst" | "rest";
}

/**
 * Create a [NestJS](https://docs.nestjs.com) TypeScript project with [Ten Mile Square](https://tenmilesquare.com)
 * opinionated defaults, and preloaded with one of a few possible sample applications.
 *
 * @pjid tms-nestjs-app
 */

export class TmsNestJSAppProject extends TmsTypeScriptAppProject {
  constructor(options: TmsNestJSAppProjectOptions) {
    const defaultOptions = {
      addDefaultBundle: false,

      esmSupportConfig: false,

      tsconfig: {
        compilerOptions: {
          module: "node16",

          // Needed for nestjs
          noImplicitAny: true,
          declaration: true,
          emitDecoratorMetadata: true,
          experimentalDecorators: true,
          allowSyntheticDefaultImports: true,
          sourceMap: true,
          outDir: "./dist",
          baseUrl: "./",

          // Let the "extends" handle these
          target: undefined,
          forceConsistentCasingInFileNames: undefined,
          strictNullChecks: undefined, // Default to true when strict: true
          noEmit: undefined,
          noFallthroughCasesInSwitch: undefined,

          // Handled below with overrides
          // removeComments: true,
          // incremental: true,
          // strictBindCallApply: true,
        },
      },

      tsconfigDev: {
        compilerOptions: {
          baseUrl: "./",

          target: "es2022",
          experimentalDecorators: undefined,
          emitDecoratorMetadata: undefined,
          allowSyntheticDefaultImports: undefined,
          outDir: undefined,
          skipLibCheck: undefined,
          strictNullChecks: undefined,
          noImplicitAny: undefined,
          forceConsistentCasingInFileNames: undefined,
          noFallthroughCasesInSwitch: undefined,
          noEmit: true,
        },
      },

      jestOptions: {
        jestConfig: {
          // testEnvironment: "node",
          moduleFileExtensions: ["js", "json", "ts"],
          rootDir: ".",
          transform: {
            "^.+\\.(t|j)s$": new javascript.Transform("ts-jest", {
              useESM: false,
              tsconfig: "tsconfig.dev.json",
            }),
          },
          // collectCoverageFrom: ["**/*.(t|j)s"],
          // coverageDirectory: "../coverage",
          testMatch: ["<rootDir>/(test|src)/**/*(*)@(.|-)@(spec|test).ts?(x)"],
          moduleNameMapper: {
            "^(\\.{1,2}/.*)\\.js$": "$1",
          },
        },
        updateSnapshot: javascript.UpdateSnapshot.NEVER,
      },

      sampleCode: true,
      sampleType: "graphql-codefirst",
    } satisfies Partial<TmsNestJSAppProjectOptions>;

    const mergedOptions = deepMerge(
      [
        defaultOptions, // will get mutated
        options,
      ],
      true,
    ) as TmsNestJSAppProjectOptions; // justified cast, since we know we're merging two of this type

    super({ ...mergedOptions, sampleCode: false });

    this.compileTask.reset("nest build");
    if (this.eslint) {
      this.addTask("format").spawn(this.eslint.eslintTask);
    }
    this.addTask("start").exec("nest start");
    this.addTask("start:dev").exec("nest start --watch");
    this.addTask("start:debug").exec("nest start --debug --watch");
    this.addTask("start:prod").exec("node dist/main");

    this.addDeps(
      "@nestjs/common@^10.0.0",
      "@nestjs/core@^10.0.0",
      "@nestjs/platform-express@^10.0.0",
      "reflect-metadata@^0.1.13",
      "rxjs@^7.8.1",

      // graphql
      "graphql",
      "@graphql-tools/utils",
      "graphql-query-complexity",
      "class-transformer",
      "class-validator",
      "graphql-subscriptions",
      "@apollo/server",
      "@nestjs/graphql",
      "@nestjs/apollo",
    );
    this.addDevDeps(
      "@tsconfig/node18",
      "@tsconfig/strictest",
      "@nestjs/cli@^10.0.0",
      "@nestjs/schematics@^10.0.0",
      "@nestjs/testing@^10.0.0",
      "@types/express@^4.17.17",
      "@types/supertest@^2.0.12",
      "supertest@^6.3.3",
      "source-map-support@^0.5.21",
      "ts-loader@^9.4.3",
      "tsconfig-paths@^4.2.0",
    );

    if (this.tsconfig) {
      // Options not yet internally supported by projen
      this.tsconfig.file.addOverride("compilerOptions.removeComments", true);
      this.tsconfig.file.addOverride("compilerOptions.incremental", true);
      this.tsconfig.file.addOverride(
        "compilerOptions.strictBindCallApply",
        true,
      );
    }

    if (mergedOptions.sampleCode) {
      if (mergedOptions.sampleType === "graphql-codefirst") {
        new SampleDir(this, this.srcdir, {
          sourceDir: path.join(
            __dirname,
            "..",
            "samples",
            "nestjs-graphql-codefirst",
            "src",
          ),
        });
        new SampleDir(this, this.testdir, {
          sourceDir: path.join(
            __dirname,
            "..",
            "samples",
            "nestjs-graphql-codefirst",
            "test",
          ),
        });
      } else if (mergedOptions.sampleType === "graphql-schemafirst") {
        // generate-typings.ts uses ts-morph, so we need to add it to devDeps
        this.addDevDeps("ts-morph");
        const generateTypingsTs = "generate-typings.ts";

        for (const fileName of [generateTypingsTs, "nest-cli.json"]) {
          new SampleFile(this, fileName, {
            sourcePath: path.join(
              __dirname,
              "..",
              "samples",
              "nestjs-graphql-schemafirst",
              "root",
              fileName,
            ),
          });
        }
        new SampleDir(this, this.srcdir, {
          sourceDir: path.join(
            __dirname,
            "..",
            "samples",
            "nestjs-graphql-schemafirst",
            "src",
          ),
        });
        new SampleDir(this, this.testdir, {
          sourceDir: path.join(
            __dirname,
            "..",
            "samples",
            "nestjs-graphql-schemafirst",
            "test",
          ),
        });

        this.eslint?.addLintPattern(generateTypingsTs);
        this.tsconfigDev.addInclude(generateTypingsTs);
        this.addTask("generate-typings", {
          exec: `ts-node --project=tsconfig.dev.json ${generateTypingsTs}`,
          description: "Generate typings from graphql files",
        });
      } else {
        throw new Error("Unsupported sample type");
      }
    }
  }
}
