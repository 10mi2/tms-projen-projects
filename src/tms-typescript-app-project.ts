import { join as joinPath, relative as relativePath, sep } from "path";
import { Component, JsonPatch, SampleDir, TextFile } from "projen";
import {
  NodePackageManager,
  TypeScriptCompilerOptions,
  TypeScriptModuleResolution,
  TypescriptConfigExtends,
  UpdateSnapshot,
} from "projen/lib/javascript";
import {
  TypeScriptAppProject,
  TypeScriptProject,
  TypeScriptProjectOptions,
} from "projen/lib/typescript";
import { deepMerge } from "projen/lib/util";

export enum TmsTSConfigBase {
  NODE_LTS = "node-lts",
  NODE18 = "node18",
  NODE20 = "node20",
}

const RESET_COMPILER_OPTIONS = {
  alwaysStrict: undefined,
  declaration: undefined,
  esModuleInterop: undefined,
  experimentalDecorators: undefined,
  inlineSourceMap: undefined,
  inlineSources: undefined,
  lib: undefined,
  module: undefined,
  noEmitOnError: undefined,
  noFallthroughCasesInSwitch: undefined,
  noImplicitAny: undefined,
  noImplicitReturns: undefined,
  noImplicitThis: undefined,
  noUnusedLocals: undefined,
  noUnusedParameters: undefined,
  resolveJsonModule: undefined,
  strict: undefined,
  strictNullChecks: undefined,
  strictPropertyInitialization: undefined,
  stripInternal: undefined,
  target: undefined,
} satisfies Partial<TypeScriptCompilerOptions>;

export interface TmsTypeScriptAppProjectOptions
  extends TypeScriptProjectOptions {
  /**
   * Add a default bundle to the project.
   *
   * Will bundle ./src/ to ./dist/ using esbuild.
   *
   * @default true
   * @featured
   */
  readonly addDefaultBundle?: boolean;

  /**
   * Configure for ESM
   *
   * @default true
   * @featured
   */
  readonly esmSupportConfig?: boolean;

  /**
   * Change the default-set eslint auto-fixable rules to "warn" instead of "error"
   *
   * @default true
   * @featured
   */
  readonly eslintFixableAsWarn?: boolean;

  /**
   * Declare a specific node version to put in `.nvmrc` for `nvm` or `fnm` to use
   *
   * NOTE: As of this writing ts-node (v10.9.1) has issues with node versions 18.19.x and newer (including 20.x)
   * when esm is enabled (`esmSupportConfig: true`)
   *
   * @default v18.18.2
   * @featured
   */
  readonly nodeVersion?: string;

  /**
   * TSConfig base configuration selection
   *
   * Using one of the options from https://github.com/tsconfig/bases as a base, then any explicit settings override
   * those. Note that only nodes18 and above are supported.
   *
   * @remarks
   *
   * Not all options are supported, perticularly those of node version that are older than is supported by Projen.
   *
   * @see {@link tsconfigBaseStrictest}
   *
   * @default TmsTSConfigBase.Node18
   *
   */
  readonly tsconfigBase?: TmsTSConfigBase;

  /**
   * TSConfig base configuration selection for `tsconfig.dev.json`, used to run projen itslef via `ts-node`
   *
   * @remarks
   *
   * Due to a bug in `ts-node` (v10.9.1) when used with Typescript 5.3.2 the tsconfig "extends" property when referring
   * to a file that's in `node_modules` will not be looked up properly. The workaround currently is to make the
   * reference relative to the tsconfig file.
   *
   * @see {@link tsconfigBase}
   *
   * @default TmsTSConfigBase.Node18
   *
   */
  readonly tsconfigBaseDev?: TmsTSConfigBase;

  /**
   * Include TSConfig "strinctest" configuration to {@link tsconfigBase}
   *
   * Using one of the options from https://github.com/tsconfig/bases as a base, then any explicit settings override
   * those. Note that only nodes18 and above are supported.
   *
   * @remarks
   *
   * If true will add the "strictest" configuration to the "extends" of the tsconfig.json file, *before* the base set by
   * the {@link tsconfigBase} option.
   *
   * Does **not** apply to `tsconfig.dev.json`.
   *
   * @see {@link tsconfigBase}
   *
   * @default true
   *
   */
  readonly tsconfigBaseStrictest?: boolean;
}

/**
 * Create a [TypeScriptAppProject](https://projen.io/api/API.html#typescriptappproject-) with
 * [Ten Mile Square](https://tenmilesquare.com) opinionated defaults.
 *
 * @pjid tms-typescript-app
 */

export class TmsTypeScriptAppProject extends TypeScriptAppProject {
  constructor(options: TmsTypeScriptAppProjectOptions) {
    const defaultOptions = {
      eslint: true,
      packageManager: NodePackageManager.NPM,
      prettier: true,
      projenrcTs: true,

      nodeVersion: "v18.18.2",

      vscode: true,
      tsconfig: {
        compilerOptions: {
          ...RESET_COMPILER_OPTIONS,
          moduleResolution: TypeScriptModuleResolution.NODE16,
          noEmit: options.addDefaultBundle ?? true,
        },
        extends: TypescriptConfigExtends.fromPaths([
          ...(options.tsconfigBaseStrictest ?? true
            ? ["@tsconfig/strictest/tsconfig.json"]
            : []),
          `@tsconfig/${
            options.tsconfigBase ?? TmsTSConfigBase.NODE18
          }/tsconfig.json`,
        ]),
      },

      tsconfigDev: {
        compilerOptions: {
          ...RESET_COMPILER_OPTIONS,
          moduleResolution: TypeScriptModuleResolution.NODE16,
          noEmit: true,
        },
        extends: TypescriptConfigExtends.fromPaths([
          `${relativePath(
            joinPath(options.outdir ?? "."),
            "./node_modules",
          ).replace(/^(?!\.)/, "./")}/@tsconfig/${
            options.tsconfigBaseDev ?? TmsTSConfigBase.NODE18
          }/tsconfig.json`,
        ]),
      },

      jest: true,
      jestOptions: {
        updateSnapshot: UpdateSnapshot.NEVER,
      },

      addDefaultBundle: true,
      esmSupportConfig: true,

      tsconfigBase: TmsTSConfigBase.NODE18,
      tsconfigBaseDev: TmsTSConfigBase.NODE18,
      tsconfigBaseStrictest: true,
    } satisfies Partial<TmsTypeScriptAppProjectOptions>;
    const mergedOptions = deepMerge(
      [
        defaultOptions, // will get mutated
        options,
      ],
      true,
    ) as TmsTypeScriptAppProjectOptions;

    super({ ...mergedOptions, sampleCode: false });

    this.addDevDeps(
      ...new Set([
        `@tsconfig/${mergedOptions.tsconfigBase ?? TmsTSConfigBase.NODE18}`,
        `@tsconfig/${mergedOptions.tsconfigBaseDev ?? TmsTSConfigBase.NODE18}`,
        ...(mergedOptions.tsconfigBaseStrictest ?? true
          ? ["@tsconfig/strictest"]
          : []),
      ]),
    );

    // Note: should adjust for https://eslint.style/guide/getting-started

    if (mergedOptions.eslintFixableAsWarn ?? true) {
      this.eslint?.addRules({
        "prettier/prettier": ["warn"],
        "import/order": [
          "warn",
          {
            groups: ["builtin", "external"],
            alphabetize: {
              order: "asc",
              caseInsensitive: true,
            },
          },
        ],
        "key-spacing": ["warn"],
        "no-multiple-empty-lines": ["warn"],
        "no-trailing-spaces": ["warn"],
        "dot-notation": ["warn"],
      });
    }

    if (mergedOptions.nodeVersion) {
      new TextFile(this, ".nvmrc", {
        lines: [mergedOptions.nodeVersion],
      });
    }

    if (mergedOptions.esmSupportConfig) {
      this.package.addField("type", "module");
      [this.tsconfig, this.tsconfigDev].forEach(
        (tsconfig) =>
          tsconfig &&
          tsconfig.file.patch(
            JsonPatch.add("/ts-node", {
              esm: true,
              preferTsExts: true,
              experimentalSpecifierResolution: "node",
            }),
          ),
      );
    } else {
      [this.tsconfig, this.tsconfigDev].forEach(
        (tsconfig) =>
          tsconfig &&
          tsconfig.file.patch(
            JsonPatch.add("/ts-node", {
              compilerOptions: {
                module: "commonjs",
              },
              preferTsExts: true,
              experimentalSpecifierResolution: "node",
            }),
          ),
      );
    }

    if (mergedOptions.jest && this.jest) {
      this.jest.config.globals = undefined;
      this.jest.config.transform = {
        "^.+\\.[tj]sx?$": [
          "ts-jest",
          {
            useESM: mergedOptions.esmSupportConfig ?? true,
            tsconfig: "tsconfig.dev.json",
          },
        ],
      };

      this.jest.config.moduleNameMapper = {
        "^(\\.{1,2}/.*)\\.js$": "$1",
      };

      if (mergedOptions.esmSupportConfig ?? true) {
        this.jest.config.preset = "ts-jest/presets/default-esm";
        this.testTask.env("NODE_OPTIONS", "--experimental-vm-modules");
      } else {
        this.jest.config.preset = "ts-jest/presets/default";
      }
    }

    if (mergedOptions.vscode) {
      if (!this.vscode?.settings) {
        throw new Error("vscode settings not found, but should have been");
      }
      if (!this.vscode?.extensions) {
        throw new Error("vscode extensions not found, but should have been");
      }

      const settings = this.vscode?.settings;
      settings.addSetting("jest.jestCommandLine", "npm test --");
      settings.addSetting("jest.rootPath", "./");

      const extensions = this.vscode?.extensions;
      extensions.addRecommendations(
        "dbaeumer.vscode-eslint",
        "Orta.vscode-jest",
      );
    }

    if (mergedOptions.addDefaultBundle ?? true) {
      this.bundler.addBundle([this.srcdir, "index.ts"].join(sep), {
        target: "node18",
        platform: "node",
        format: mergedOptions.esmSupportConfig ?? true ? "esm" : "cjs",
        sourcemap: true,
      });
    }
    if (mergedOptions.sampleCode ?? true) {
      new SampleCode(this);
    }
  }
}

class SampleCode extends Component {
  constructor(project: TypeScriptProject) {
    super(project);
    const indexSrcCode = [
      'import { Hello } from "./hello.js";',
      "",
      "console.log(await new Hello().sayHello(2000));",
    ].join("\n");
    const helloSrcCode = [
      'import { setTimeout } from "timers/promises";',
      "export class Hello {",
      "  public async sayHello(delay: number = 100): Promise<string> {",
      "    await setTimeout(delay);",
      '    return "hello, world!";',
      "  }",
      "}",
    ].join("\n");

    const testCode = [
      'import { Hello } from "../src/hello.js";',
      "",
      'test("hello", async () => {',
      "  const hello = new Hello();",
      '  expect(await hello.sayHello()).toBe("hello, world!");',
      "});",
    ].join("\n");

    new SampleDir(project, project.srcdir, {
      files: {
        "index.ts": indexSrcCode,
        "hello.ts": helloSrcCode,
      },
    });

    if (project.jest) {
      new SampleDir(project, project.testdir, {
        files: {
          "hello.test.ts": testCode,
        },
      });
    }
  }
}
