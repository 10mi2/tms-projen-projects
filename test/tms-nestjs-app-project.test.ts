import { Testing } from "projen/lib/testing";
import { TmsNestJSAppProject } from "../src";

const originalProcess = process;
beforeEach(() => {
  global.process = {
    ...originalProcess,
    versions: { ...originalProcess.versions, node: "18.18.2" },
  };
});
afterEach(() => {
  global.process = originalProcess;
});

test("TmsNestJSAppProject has reasonable configuration", () => {
  const project = new TmsNestJSAppProject({
    name: "test",
    defaultReleaseBranch: "main",
    // default settings
  });
  const snapshot = Testing.synth(project);

  expect(snapshot["tsconfig.json"]).toMatchSnapshot();
  expect(snapshot["tsconfig.dev.json"]).toMatchSnapshot();
  expect(snapshot["package.json"]).toMatchSnapshot();
  expect(snapshot[".vscode/settings.json"]).toMatchSnapshot();
  expect(snapshot[".vscode/extensions.json"]).toMatchSnapshot();

  expect(snapshot[".eslintrc.json"]).toMatchSnapshot();

  expect(snapshot["src/main.ts"]).toBeDefined();
  expect(snapshot["test/app.e2e-spec.ts"]).toBeDefined();
});

test("TmsNestJSAppProject doesn't make sample code when told not to", () => {
  const project = new TmsNestJSAppProject({
    name: "test",
    defaultReleaseBranch: "main",
    sampleCode: false,
    // default settings
  });
  const snapshot = Testing.synth(project); // synthSnapshot(project, { parseJson: true });

  expect(snapshot["src/main.ts"]).toBeUndefined();
  expect(snapshot["test/app.e2e-spec.ts"]).toBeUndefined();
});

test("TmsNestJSAppProject makes schemafirst sample code when told not to", () => {
  const project = new TmsNestJSAppProject({
    name: "test",
    defaultReleaseBranch: "main",
    sampleType: "graphql-schemafirst",
    // sampleCode: false,
    // default settings
  });
  const snapshot = Testing.synth(project); // synthSnapshot(project, { parseJson: true });

  expect(snapshot["src/main.ts"]).toBeDefined();
  expect(snapshot["test/app.e2e-spec.ts"]).toBeDefined();
  expect(snapshot["src/graphql.schema.ts"]).toBeDefined();
  expect(snapshot["generate-typings.ts"]).toBeDefined();
  expect(snapshot["nest-cli.json"]).toBeDefined();
});
