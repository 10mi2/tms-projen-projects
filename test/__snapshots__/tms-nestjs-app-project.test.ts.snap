// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TmsNestJSAppProject has reasonable configuration 1`] = `
{
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "baseUrl": "./",
    "checkJs": true,
    "declaration": true,
    "emitDecoratorMetadata": true,
    "esModuleInterop": true,
    "exactOptionalPropertyTypes": true,
    "experimentalDecorators": true,
    "incremental": true,
    "isolatedModules": true,
    "module": "node16",
    "moduleResolution": "node16",
    "noFallthroughCasesInSwitch": true,
    "noImplicitAny": true,
    "noImplicitOverride": true,
    "noImplicitReturns": true,
    "noPropertyAccessFromIndexSignature": true,
    "noUncheckedIndexedAccess": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "outDir": "./dist",
    "removeComments": true,
    "rootDir": "src",
    "skipLibCheck": true,
    "sourceMap": true,
    "strict": true,
    "strictBindCallApply": true,
    "strictNullChecks": true,
    "target": "ES2019",
  },
  "exclude": [],
  "extends": "@tsconfig/node18/tsconfig.json",
  "include": [
    "src/**/*.ts",
  ],
  "ts-node": {
    "compilerOptions": {
      "module": "commonjs",
    },
    "experimentalSpecifierResolution": "node",
    "preferTsExts": true,
  },
}
`;

exports[`TmsNestJSAppProject has reasonable configuration 2`] = `
{
  "compilerOptions": {
    "allowSyntheticDefaultImports": true,
    "baseUrl": "./",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "moduleResolution": "node16",
    "noEmit": true,
    "noFallthroughCasesInSwitch": true,
    "noImplicitAny": true,
    "outDir": "./dist",
    "skipLibCheck": true,
    "sourceMap": true,
    "strictNullChecks": true,
    "target": "es2022",
  },
  "exclude": [
    "node_modules",
  ],
  "extends": "@tsconfig/node18/tsconfig.json",
  "include": [
    "src/**/*.ts",
    "test/**/*.ts",
    ".projenrc.ts",
    "projenrc/**/*.ts",
  ],
  "ts-node": {
    "compilerOptions": {
      "module": "commonjs",
    },
    "experimentalSpecifierResolution": "node",
    "preferTsExts": true,
  },
}
`;

exports[`TmsNestJSAppProject has reasonable configuration 3`] = `
{
  "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
  "dependencies": {
    "@apollo/server": "*",
    "@graphql-tools/utils": "*",
    "@nestjs/apollo": "*",
    "@nestjs/common": "^10.0.0",
    "@nestjs/core": "^10.0.0",
    "@nestjs/graphql": "*",
    "@nestjs/platform-express": "^10.0.0",
    "class-transformer": "*",
    "class-validator": "*",
    "graphql": "*",
    "graphql-query-complexity": "*",
    "graphql-subscriptions": "*",
    "reflect-metadata": "^0.1.13",
    "rxjs": "^7.8.1",
  },
  "devDependencies": {
    "@nestjs/cli": "^10.0.0",
    "@nestjs/schematics": "^10.0.0",
    "@nestjs/testing": "^10.0.0",
    "@tsconfig/node18": "*",
    "@tsconfig/strictest": "*",
    "@types/express": "^4.17.17",
    "@types/jest": "*",
    "@types/node": "^18",
    "@types/supertest": "^2.0.12",
    "@typescript-eslint/eslint-plugin": "^6",
    "@typescript-eslint/parser": "^6",
    "constructs": "^10.0.0",
    "eslint": "^8",
    "eslint-config-prettier": "*",
    "eslint-import-resolver-typescript": "*",
    "eslint-plugin-import": "*",
    "eslint-plugin-prettier": "*",
    "jest": "*",
    "jest-junit": "^15",
    "prettier": "*",
    "projen": "*",
    "source-map-support": "^0.5.21",
    "supertest": "^6.3.3",
    "ts-jest": "*",
    "ts-loader": "^9.4.3",
    "ts-node": ">=10.9.2",
    "tsconfig-paths": "^4.2.0",
    "typescript": "*",
  },
  "jest": {
    "clearMocks": true,
    "collectCoverage": true,
    "coverageDirectory": "coverage",
    "coveragePathIgnorePatterns": [
      "/node_modules/",
    ],
    "coverageReporters": [
      "json",
      "lcov",
      "clover",
      "cobertura",
      "text",
    ],
    "moduleFileExtensions": [
      "js",
      "json",
      "ts",
    ],
    "moduleNameMapper": {
      "^(\\.{1,2}/.*)\\.m?js$": "$1",
    },
    "reporters": [
      "default",
      [
        "jest-junit",
        {
          "outputDirectory": "test-reports",
        },
      ],
    ],
    "rootDir": ".",
    "testMatch": [
      "<rootDir>/(test|src)/**/*(*)@(.|-)@(spec|test).ts?(x)",
      "<rootDir>/src/**/__tests__/**/*.ts?(x)",
      "<rootDir>/(test|src)/**/*(*.)@(spec|test).ts?(x)",
    ],
    "testPathIgnorePatterns": [
      "/node_modules/",
    ],
    "transform": {
      "^.+\\.m?[tj]sx?$": [
        "ts-jest",
        {
          "tsconfig": "tsconfig.dev.json",
          "useESM": false,
        },
      ],
    },
    "watchPathIgnorePatterns": [
      "/node_modules/",
    ],
  },
  "license": "Apache-2.0",
  "name": "test",
  "scripts": {
    "build": "npx projen build",
    "clobber": "npx projen clobber",
    "compile": "npx projen compile",
    "default": "npx projen default",
    "eject": "npx projen eject",
    "eslint": "npx projen eslint",
    "format": "npx projen format",
    "package": "npx projen package",
    "post-compile": "npx projen post-compile",
    "post-upgrade": "npx projen post-upgrade",
    "pre-compile": "npx projen pre-compile",
    "projen": "npx projen",
    "start": "npx projen start",
    "start:debug": "npx projen start:debug",
    "start:dev": "npx projen start:dev",
    "start:prod": "npx projen start:prod",
    "test": "npx projen test",
    "test:update": "npx projen test:update",
    "test:watch": "npx projen test:watch",
    "upgrade": "npx projen upgrade",
    "watch": "npx projen watch",
  },
  "version": "0.0.0",
}
`;

exports[`TmsNestJSAppProject has reasonable configuration 4`] = `
{
  "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
  "jest.jestCommandLine": "npm test --",
  "jest.rootPath": "./",
}
`;

exports[`TmsNestJSAppProject has reasonable configuration 5`] = `
{
  "recommendations": [
    "dbaeumer.vscode-eslint",
    "Orta.vscode-jest",
  ],
}
`;

exports[`TmsNestJSAppProject has reasonable configuration 6`] = `
{
  "env": {
    "jest": true,
    "node": true,
  },
  "extends": [
    "plugin:import/typescript",
    "plugin:prettier/recommended",
  ],
  "ignorePatterns": [
    "*.js",
    "*.d.ts",
    "node_modules/",
    "*.generated.ts",
    "coverage",
    "!.projenrc.ts",
    "!projenrc/**/*.ts",
  ],
  "overrides": [
    {
      "files": [
        ".projenrc.ts",
      ],
      "rules": {
        "@typescript-eslint/no-require-imports": "off",
        "import/no-extraneous-dependencies": "off",
      },
    },
  ],
  "parser": "@typescript-eslint/parser",
  "parserOptions": {
    "ecmaVersion": 2018,
    "project": "./tsconfig.dev.json",
    "sourceType": "module",
  },
  "plugins": [
    "@typescript-eslint",
    "import",
  ],
  "root": true,
  "rules": {
    "@typescript-eslint/member-ordering": [
      "error",
      {
        "default": [
          "public-static-field",
          "public-static-method",
          "protected-static-field",
          "protected-static-method",
          "private-static-field",
          "private-static-method",
          "field",
          "constructor",
          "method",
        ],
      },
    ],
    "@typescript-eslint/no-floating-promises": [
      "error",
    ],
    "@typescript-eslint/no-require-imports": [
      "error",
    ],
    "@typescript-eslint/no-shadow": [
      "error",
    ],
    "@typescript-eslint/return-await": [
      "error",
    ],
    "dot-notation": [
      "warn",
    ],
    "import/no-extraneous-dependencies": [
      "error",
      {
        "devDependencies": [
          "**/test/**",
          "**/build-tools/**",
          ".projenrc.ts",
          "projenrc/**/*.ts",
        ],
        "optionalDependencies": false,
        "peerDependencies": true,
      },
    ],
    "import/no-unresolved": [
      "error",
    ],
    "import/order": [
      "warn",
      {
        "alphabetize": {
          "caseInsensitive": true,
          "order": "asc",
        },
        "groups": [
          "builtin",
          "external",
        ],
      },
    ],
    "key-spacing": [
      "warn",
    ],
    "no-bitwise": [
      "error",
    ],
    "no-duplicate-imports": [
      "error",
    ],
    "no-multiple-empty-lines": [
      "warn",
    ],
    "no-return-await": [
      "off",
    ],
    "no-shadow": [
      "off",
    ],
    "no-trailing-spaces": [
      "warn",
    ],
    "prettier/prettier": [
      "warn",
    ],
  },
  "settings": {
    "import/parsers": {
      "@typescript-eslint/parser": [
        ".ts",
        ".tsx",
      ],
    },
    "import/resolver": {
      "node": {},
      "typescript": {
        "alwaysTryTypes": true,
        "project": "./tsconfig.dev.json",
      },
    },
  },
}
`;
